src/uct/tcp/tcp_ep.c

```diff
@@ -290,15 +298,44 @@ void uct_tcp_ep_destroy(uct_ep_h tl_ep)
     }
 }
 
-void uct_tcp_ep_set_failed(uct_tcp_ep_t *ep)
+static unsigned uct_tcp_ep_failed_progress(void *arg)
 {
+    uct_tcp_ep_t *ep       = (uct_tcp_ep_t*)arg;
     uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
                                             uct_tcp_iface_t);
 
+    ucs_assert(ep->flags & UCT_TCP_EP_FLAG_FAILED);
+
+    ep->failed_cb_id = UCS_CALLBACKQ_ID_NULL;
+
     uct_tcp_cm_change_conn_state(ep, UCT_TCP_EP_CONN_STATE_CLOSED);
     uct_set_ep_failed(&UCS_CLASS_NAME(uct_tcp_ep_t),
                       &ep->super.super, &iface->super.super,
                       UCS_ERR_ENDPOINT_TIMEOUT);
+
+    return 1;
+}
+
+void uct_tcp_ep_set_failed(uct_tcp_ep_t *ep)
+{
+    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                            uct_tcp_iface_t);
+
+    if (ep->flags & UCT_TCP_EP_FLAG_FAILED) {
+        /* error handling callback has already been scheduled and socket
+         * fd has been closed, no need do to these action one more time */
+        ucs_assert(ep->failed_cb_id != UCS_CALLBACKQ_ID_NULL);
+        return;
+    }
+
+    uct_tcp_ep_mod_events(ep, 0, ep->events);
+    ucs_close_fd(&ep->fd);
+    ep->flags        |= UCT_TCP_EP_FLAG_FAILED;
+    ep->failed_cb_id  = UCS_CALLBACKQ_ID_NULL;
```

Comment:
but ucs_close_fd calls ucs_close_fd() again right? from uct_set_ep_failed()
i think it works just because UCS_CALLBACKQ_ID_NULL is -1
IMO we should not save ep->failed_cb_id at all, but use filter-remove to remove it from cbq during destroy flow

Reply:
good catch, didn't know about this
done